apps/openmw/mwclass/npc.cpp

```diff
@@ -1414,8 +1414,19 @@ namespace MWClass
         return ref->mBase->mFaction;
     }
 
-    int Npc::getPrimaryFactionRank (const MWWorld::ConstPtr& ptr) const
+    int Npc::getPrimaryFactionRank (const MWWorld::Ptr& ptr) const
     {
+        std::string factionID = ptr.getClass().getPrimaryFaction(ptr);
+        if(factionID.empty())
+            return -1;
+
+        // Search in the NPC data first
```

Comment:
The added getFactionRank method looks pretty much identical to the code following this line. I think something like 
```cpp
int rank = ptr.getClass().getNpcStats(ptr).getFactionRank();
if (rank != -1)
    return rank;
```
should be possible to do.

Or, to make getPrimaryFactionRank const-correct again (accept a ConstPtr), we can really assume NPCs without custom data (obviously) never have a faction ranks map set up, so it's also possible to do something like this:
```cpp
if (const MWWorld::CustomData* data = ptr.getRefData().getCustomData())
{
    int rank = data->asNpcCustomData().mNpcStats.getFactionRank();
    if (rank >= 0)
        return rank;
}
```

Reply:
> int rank = ptr.getClass().getNpcStats(ptr).getFactionRank();

I do not understand you. There is no getFactionRank() function without arguments in the NpcStats.

Reply:
Using the getFactionRank(factionID) now.